src/services/announcement-service.ts

```diff
@@ -0,0 +1,206 @@
+    }).then((d) => d.data)
+      .catch(() => undefined);
+    //TODO: handle refresh/ 401/403
+  }
+
+  static async buildAnnouncementEmbed(announcement: CanvasAnnouncement, courseID: string, canvasInstanceID: string, discordUserID: string): Promise<MessageEmbed> {
+    const ts = new TurndownService();
+
+    const courses = await CanvasController.getCourses(discordUserID, canvasInstanceID);
+    if (courses === undefined) {
+      throw new Error('Courses not defined. Likely invalid or undefined token from users.');
+    }
+    const course = courses.find(c => c.id === parseInt(courseID));
+
+    const postedTime = new Date(announcement.posted_at);
+    const postTimeString = postedTime.getHours() + ':' + (postedTime.getMinutes() < 10 ? '0' + postedTime.getMinutes() : postedTime.getMinutes()) + ' • ' + postedTime.getDate() + '/' + (postedTime.getMonth() + 1) + '/' + postedTime.getFullYear();
+
+    const embed = new MessageEmbed({
+      color: '#E63F30',
+      title: announcement.title,
+      url: announcement.html_url,
+      author: {
+        name: course?.course_code + ' - ' + announcement.author.display_name
+      },
+      description: ts.turndown(announcement.message),
+      footer: {
+        text: postTimeString
+      }
+    });
+    return embed;
+  }
+
+  // ## Checking announcements for changes
+  // FIX: API checks before bot can receive events -> lastAnnounce is set to something that isn't posted
+  // TODO: check rate limits. Currently 1 minute interval. We want this as low as is allowed.
+  // guildID: string, CanvasInstanceID: string
+  static initAnnouncementJob(): NodeJS.Timeout {
+    return setInterval(async () => {
+      const guilds = (await db.collection(Collections.guilds).get()).docs.map((d) => d.data());
+      // FIX: This for loop can prob be done better or differently.
+      for (let i = 0; i < guilds.length; i++) {
+        const guildID = guilds[i]?.id;
+        const canvasInstanceID = guilds[i]?.canvasInstanceID;
+
+        if (guildID === undefined || canvasInstanceID === undefined) {
+          continue;
+        }
+        const canvas = (await db.collection(Collections.canvas).doc(canvasInstanceID).get()).data();
+        /*change => is just guild in loop */
+        const guildConfig = (await db.collection(Collections.guilds).doc(guildID).get()).data();
+        if (guildConfig === undefined) {
+          console.error(`Guildconfig with id ${guildID} is undefined.`);
+          return;
+        }
+        if (canvas === undefined) {
+          console.error(`CanvasInstance with id ${canvasInstanceID} is undefined.`);
+          return;
+        }
+        console.log('Canvas domain: ', canvas.endpoint);
+
+        // Canvas Instance for announcements is undefined or empty
+        // TODO: there is probably a better way of checking if it's empty
+        if (canvas.lastAnnounce === undefined || (stringify(canvas.lastAnnounce) === stringify({}))) {
+          canvas.lastAnnounce = {};
+        }
```

Comment:
no need to check if object is empty
if you do need to check if an object is empty use `Object.keys(obj).length === 0`

use permissive null check instead of hard undefined
`canvas.lastAnnounce == null` checks if object is null or undefined
`canvas.lastAnnounce === undefined` only checks if object is undefined

Reply:
Resolved in #21